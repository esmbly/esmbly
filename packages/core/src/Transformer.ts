import {
  File,
  Output,
  OutputFormat,
  SyntaxTree,
  TransformerInterface,
} from '@esmbly/types';
import { fileTypeForOutputFormat } from '@esmbly/utils';

export abstract class Transformer implements TransformerInterface {
  public static outputFormats: OutputFormat[] = [];

  public abstract transform(syntaxTrees: SyntaxTree[]): void;

  public createFiles(trees: SyntaxTree[], output: Output[]): File[] {
    const files: File[] = [];
    trees.forEach((tree: SyntaxTree) => {
      output.forEach(({ dir, format }: Output) => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        if ((this.constructor as any).outputFormats.includes(format)) {
          files.push({
            ...tree.represents,
            content: tree.toCode(),
            dir: dir || tree.represents.dir,
            type: fileTypeForOutputFormat(format),
          });
        }
      });
    });
    return files;
  }
}
