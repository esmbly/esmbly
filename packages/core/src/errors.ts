export const MissingConfig = (): string => {
  return `Please provide a valid configuration object.
    {
      input: File[];
      transformers: Transformer[];
      output: Output[];
    }
  `;
};

export const NoInput = (): string => {
  return 'Please specify at least one file to transform.';
};

export const NoTransformers = (): string => {
  return 'Please specify at least one transformer.';
};

export const NoOutput = (): string => {
  return 'Please specify at least one output format.';
};
