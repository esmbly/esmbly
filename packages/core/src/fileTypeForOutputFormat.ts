import { FileType, Format } from '@esmbly/types';

export function fileTypeForOutputFormat(format: Format): FileType {
  switch (format) {
    case Format.Asm:
      return FileType.Asm;
    case Format.Flow:
      return FileType.JavaScript;
    case Format.TypeScript:
      return FileType.TypeScript;
    case Format.Wat:
      return FileType.Wat;
    case Format.WebAssembly:
      return FileType.WebAssembly;
    case Format.AssemblyScript:
      return FileType.TypeScript;
    default:
      throw new Error(`Output format: ${format} is not supported`);
  }
}
