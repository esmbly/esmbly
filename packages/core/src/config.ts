import { RunConfig, Transformer, OutputFormat, Output } from '@esmbly/types';
import * as errors from './errors';

export function validateRunConfig(config: RunConfig): void {
  if (!config || !config.input || !config.transformers || !config.output) {
    throw new Error(errors.MissingConfig());
  }
  if (config.input.length < 1) {
    throw new Error(errors.NoInput());
  }
  if (config.transformers.length < 1) {
    throw new Error(errors.NoTransformers());
  }
  if (config.output.length < 1) {
    throw new Error(errors.NoOutput());
  }
  config.transformers.forEach((transformer: Transformer) => {
    if (typeof transformer.run !== 'function') {
      throw new Error(errors.InvalidTransformer(transformer));
    }
  });
  config.output.forEach((out: Output) => {
    let format = typeof out === 'string' ? out : out.format;
    if (format in OutputFormat === false) {
      throw new Error(errors.InvalidOutput(format));
    }
  });
}
