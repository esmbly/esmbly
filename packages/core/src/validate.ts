import { Format, RunConfig, SyntaxTree, Transformer } from '@esmbly/types';
import * as errors from './errors';

export function validateConfig(config?: RunConfig): void {
  if (!config || !config.input || !config.transformers || !config.output) {
    throw new Error(errors.MissingConfig());
  }
  if (config.input.length < 1) {
    throw new Error(errors.NoInput());
  }
  if (config.transformers.length < 1) {
    throw new Error(errors.NoTransformers());
  }
  if (config.output.length < 1) {
    throw new Error(errors.NoOutput());
  }
}

export function validateInputFormat(
  trees: SyntaxTree[],
  transformer: Transformer,
): void {
  trees.forEach((tree: SyntaxTree) => {
    if (transformer.format.input === Format.Any) {
      return;
    }
    if (tree.format !== transformer.format.input) {
      const { name } = transformer;
      throw new Error(
        `Transformer: ${name} does not support input format ${tree.format}`,
      );
    }
  });
}
