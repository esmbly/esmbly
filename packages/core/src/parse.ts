import * as parser from '@babel/parser';
import { File, Format, Output, SyntaxTree, Transformer } from '@esmbly/types';
import recast from 'recast';
import { fileTypeForOutputFormat } from './fileTypeForOutputFormat';

export function parse(files: File[], transformer: Transformer): SyntaxTree[] {
  return files.map(
    (file: File): SyntaxTree => {
      try {
        const tree = recast.parse(file.content.toString(), {
          parser: transformer.parser || {
            parse(source: string) {
              return parser.parse(source, {
                plugins: transformer.parserPlugins || [],
                sourceType: 'module',
              });
            },
          },
        });
        return {
          format: transformer.format.input,
          represents: file,
          setFormat(format: Format): void {
            this.format = format;
            this.represents.type = fileTypeForOutputFormat(format);
          },
          toCode(): string {
            return recast.print(tree).code;
          },
          toFile(output: Output, content?: string | Buffer): File {
            return {
              ...this.represents,
              content: content || this.toCode(),
              outputOptions: output,
            };
          },
          tree,
        };
      } catch (err) {
        const filePath = `${file.dir}/${file.name}`;
        throw new Error(
          `Error while parsing file: ${filePath}. ${err.toString()}`,
        );
      }
    },
  );
}
