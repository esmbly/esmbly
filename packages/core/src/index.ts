import { File, RunConfig } from '@esmbly/types';
import { parse } from './parse';
import { validateConfig, validateInputFormat } from './validate';

export async function run(config: RunConfig): Promise<File[]> {
  validateConfig(config);

  let files: File[] = [];
  const trees = parse(config.input, config.transformers[0]);

  for (const transformer of config.transformers) {
    if (transformer.before) {
      await transformer.before();
    }

    validateInputFormat(trees, transformer);

    if (transformer.transform) {
      await transformer.transform(trees);
    }

    if (transformer.createFiles) {
      files = [
        ...files,
        ...(await transformer.createFiles(trees, config.output)),
      ];
    }

    if (transformer.after) {
      await transformer.after();
    }
  }

  return files;
}
