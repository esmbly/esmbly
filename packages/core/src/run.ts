import { File, RunConfig, Transformer } from '@esmbly/types';
import parse from '@esmbly/parser';
import { validateRunConfig } from './config';

export default async function run({
  input,
  transformers,
  output,
}: RunConfig): Promise<File[]> {
  validateRunConfig({ input, output, transformers });
  let files: File[] = [];
  const trees = parse(input);
  transformers.forEach((transformer: Transformer) => {
    transformer.transform(trees);
    files = [...files, ...transformer.createFiles(trees, output)];
  });
  return files;
}
