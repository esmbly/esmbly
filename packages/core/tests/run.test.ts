import * as parser from '@esmbly/parser';
import { RunConfig } from '@esmbly/types';
import run from '../src/run';
import mockAst from './__fixtures__/trees';
import mockConfig, { MockTransformer } from './__fixtures__/config';
import * as config from '../src/config';

jest.mock('../src/config');

describe('run', () => {
  it('validates the provided config', async () => {
    const configSpy = jest.spyOn(config, 'validateRunConfig');
    const parserSpy = jest.spyOn(parser, 'default');
    parserSpy.mockReturnValue(mockAst);
    configSpy.mockReturnValue();
    await run(mockConfig);
    expect(configSpy).toHaveBeenCalledTimes(1);
    expect(configSpy).toHaveBeenCalledWith(mockConfig);
    configSpy.mockRestore();
    parserSpy.mockRestore();
  });

  it('parses the provided files', async () => {
    const configSpy = jest.spyOn(config, 'validateRunConfig');
    const parserSpy = jest.spyOn(parser, 'default');
    parserSpy.mockReturnValue(mockAst);
    configSpy.mockReturnValue();
    await run(mockConfig);
    expect(parserSpy).toHaveBeenCalledTimes(1);
    expect(parserSpy).toHaveBeenCalledWith(mockConfig.input);
    configSpy.mockRestore();
    parserSpy.mockRestore();
  });

  it('passes the ast to each transformer', async () => {
    const configSpy = jest.spyOn(config, 'validateRunConfig');
    const parserSpy = jest.spyOn(parser, 'default');
    parserSpy.mockReturnValue(mockAst);
    configSpy.mockReturnValue();
    const transformerA = new MockTransformer();
    const transformerB = new MockTransformer();
    const transformSpyA = jest.spyOn(transformerA, 'transform');
    const transformSpyB = jest.spyOn(transformerB, 'transform');
    const runConfig: RunConfig = {
      ...mockConfig,
      transformers: [transformerA, transformerB],
    };
    await run(runConfig);
    expect(transformSpyA).toHaveBeenCalledTimes(1);
    expect(transformSpyB).toHaveBeenCalledWith(mockAst);
    expect(transformSpyA).toHaveBeenCalledTimes(1);
    expect(transformSpyB).toHaveBeenCalledWith(mockAst);
    configSpy.mockRestore();
    parserSpy.mockRestore();
  });

  it('returns an array of files to be outputted', async () => {
    const configSpy = jest.spyOn(config, 'validateRunConfig');
    const parserSpy = jest.spyOn(parser, 'default');
    parserSpy.mockReturnValue(mockAst);
    configSpy.mockReturnValue();
    const output = await run(mockConfig);
    expect(output).toMatchSnapshot();
    configSpy.mockRestore();
    parserSpy.mockRestore();
  });
});
