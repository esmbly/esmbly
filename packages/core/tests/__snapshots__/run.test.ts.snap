// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run returns an array of files to be outputted 1`] = `
Array [
  Object {
    "content": "function bar() { return 1; }",
    "dir": "/",
    "name": "fileA",
    "outputOptions": Object {
      "format": "Flow",
    },
    "type": ".js",
  },
  Object {
    "content": "function test() { return 2; }",
    "dir": "/dir",
    "name": "fileB",
    "outputOptions": Object {
      "format": "Flow",
    },
    "type": ".js",
  },
]
`;
